libp2p/NodeTable.cpp

```diff
@@ -674,6 +685,69 @@ std::shared_ptr<NodeEntry> NodeTable::handlePingNode(
     return sourceNodeEntry;
 }
 
+std::shared_ptr<NodeEntry> NodeTable::handleENRRequest(
+    bi::udp::endpoint const& _from, DiscoveryDatagram const& _packet)
+{
+    std::shared_ptr<NodeEntry> sourceNodeEntry = nodeEntry(_packet.sourceid);
+    if (!sourceNodeEntry)
+    {
+        LOG(m_logger) << "Source node (" << _packet.sourceid << "@" << _from
+                      << ") not found in node table. Ignoring ENRRequest request.";
+        return {};
+    }
+    if (sourceNodeEntry->endpoint() != _from)
+    {
+        LOG(m_logger) << "ENRRequest packet from unexpected endpoint " << _from << " instead of "
+                      << sourceNodeEntry->endpoint();
+        return {};
+    }
+    if (!sourceNodeEntry->lastPongReceivedTime)
+    {
+        LOG(m_logger) << "Unexpected ENRRequest packet! Endpoint proof hasn't been performed yet.";
+        return {};
+    }
+    if (!sourceNodeEntry->hasValidEndpointProof())
+    {
+        LOG(m_logger) << "Unexpected ENRRequest packet! Endpoint proof has expired.";
+        return {};
+    }
+
+    auto const& in = dynamic_cast<ENRRequest const&>(_packet);
+
+    ENRResponse response(_from, m_hostENR);
+    LOG(m_logger) << response.typeName() << " to " << in.sourceid << "@" << _from;
+    response.timestamp = nextRequestExpirationTime();
+    response.echo = in.echo;
+    response.sign(m_secret);
+    m_socket->send(response);
+
+    return sourceNodeEntry;
+}
+
+std::shared_ptr<NodeEntry> NodeTable::handleENRResponse(
```

Comment:
We should verify the hash in the response against the request (or are you planning on doing that in another change)?

Reply:
Yes, I planned to add the validation when implementing doing something useful with these responses (the last "ENR Client" part in https://github.com/ethereum/aleth/issues/5551)
But now I also think there's not much useful we can actually do with them in v4 discovery, so maybe I'll leave it just ignoring ENRResponses

Reply:
If we've enabled the functionality I think we should validate the responses so that way there's 1 less thing to do when we actually do something useful with them.